### PR TITLE
ContextManager

### LINKED ISSUE #2: Create DatabaseManager
Think of a way to manage mongo database and collection connections

### PR DESCRIPTION
**Pull request by visschersm**

### CODE REVIEW

File: src/Utilities.Mongo/Interfaces/IMongoContext.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+﻿using MongoDB.Driver;
+using System.Threading.Tasks;
+
+namespace MatrTech.Utilities.Mongo.Interfaces
+{
+    public interface IMongoContext
+    {
+        /// <summary>
+        /// Checks to see if a collection exists within the context.
```
Comment (pjotrcasteel) [inline src/Utilities.Mongo/Interfaces/IMongoContext.cs]: Checks whether a collection with the name <param name="collectionName"> exists within the context.

File: src/Utilities.Mongo/Models/MongoContext.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+#else
+                    if (genericType is not null && !collections.Contains(genericType))
+#endif
+                    {
+                        collections.Add(genericType);
+                        var method = typeof(IMongoDatabase).GetMethod(nameof(IMongoDatabase.GetCollection)) ?? throw new Exception();
+                        method = method.MakeGenericMethod(genericType) ?? throw new Exception();
+
+                        // TODO: verify name or get it from configuration
+                        var collectionName = propertyInfo.Name.Replace("Collection", "");
```
Comment (pjotrcasteel) [inline src/Utilities.Mongo/Models/MongoContext.cs]: replace ("collection", string.empty)

File: tests/Utilities.Mongo.UnitTests/ContextManagerTests.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    public class ContextManagerTests
+    {
+        private const string connectionUrl = "mongodb://localhost:27017";
+
+        [TestMethod]
+        public void Create_DerivedFromMongoContext_ShouldNotBeNull()
+        {
+            var databaseName = $"{Guid.NewGuid()}";
+            ContextManager.Create<TestContext>(connectionUrl, databaseName)
+                .Should().NotBeNull();
```
Comment (pjotrcasteel) [inline tests/Utilities.Mongo.UnitTests/ContextManagerTests.cs]: check (line)formatting in file

File: src/Utilities.Mongo/ContextManager.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
+﻿using MatrTech.Utilities.Helpers;
+using MatrTech.Utilities.Mongo.Models;
+using MongoDB.Driver;
+using System;
+
+namespace MatrTech.Utilities.Mongo
+{
+    public class ContextManager
+    {
+        public static TContext Create<TContext>(string connectionUrl, string databaseName)
```
Comment (pjotrcasteel) [inline src/Utilities.Mongo/ContextManager.cs:20]: missing summary.

File: src/Utilities.Mongo/Interfaces/IMongoContext.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
+﻿using MongoDB.Driver;
+using System.Threading.Tasks;
+
+namespace MatrTech.Utilities.Mongo.Interfaces
+{
+    public interface IMongoContext
```
Comment (pjotrcasteel) [inline src/Utilities.Mongo/Interfaces/IMongoContext.cs:6]: I wonder if the summary of the interface is replaced by the implemantion class.

File: src/Utilities.Mongo/Models/MongoContext.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+using System.ComponentModel;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace MatrTech.Utilities.Mongo.Models
+{
+    public abstract class MongoContext : IMongoContext
+    {
+        private readonly IMongoDatabase database;
+        private readonly IList<Type> collections = new List<Type>();
```
Comment (pjotrcasteel) [inline src/Utilities.Mongo/Models/MongoContext.cs:14]: how does one add collections

File: tests/Utilities.Mongo.UnitTests/ContextManagerTests.cs (hunk 3 above)
Comment (visschersm) [reply tests/Utilities.Mongo.UnitTests/ContextManagerTests.cs]: What do you mean by this?

Comment (pjotrcasteel) [reply tests/Utilities.Mongo.UnitTests/ContextManagerTests.cs]: ...).Should().NotBeNull();

vs.

...)
.Should()
.NotBeNull();


consistancy in file

File: tests/Utilities.Mongo.UnitTests/ContextManagerTests.cs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        }
+
+        [TestMethod]
+        public void Create_GetCollectionAfterCreate_ShouldNotBeNull()
+        {
+            var databaseName = $"{Guid.NewGuid()}";
+            var context = ContextManager.Create<TestContext>(connectionUrl, databaseName);
+            string collectionName = $"test-{Guid.NewGuid()}";
```
Comment (pjotrcasteel) [inline tests/Utilities.Mongo.UnitTests/ContextManagerTests.cs]: this variable doens't do anything right?

File: src/Utilities.Mongo/Models/MongoContext.cs (hunk 6 above)
Comment (visschersm) [reply src/Utilities.Mongo/Models/MongoContext.cs:14]: Will resolve this in issue: #10

File: src/Utilities.Mongo/Models/MongoContext.cs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                    if (!(genericType is null) && !collections.Contains(genericType))
+#else
+                    if (genericType is not null && !collections.Contains(genericType))
+#endif
+                    {
+                        collections.Add(genericType);
+                        var method = typeof(IMongoDatabase).GetMethod(nameof(IMongoDatabase.GetCollection)) ?? throw new Exception();
+                        method = method.MakeGenericMethod(genericType) ?? throw new Exception();
+
+                        // TODO: verify name or get it from configuration
```
Comment (visschersm) [inline src/Utilities.Mongo/Models/MongoContext.cs]: Remove TODO and resolve in issue: #1

File: src/Utilities.Mongo/Utilities.Mongo.csproj [hunk 9]
```diff
@@ -9,16 +9,22 @@
...
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
     <WarningsAsErrors />
- 
+
     <MajorVersion>0</MajorVersion>
     <MinorVersion>0</MinorVersion>
     <PatchVersion>1</PatchVersion>
     <BuildVersion>$([System.DateTime]::Now.ToString("yyMMddHHmm"))</BuildVersion>
     <Version>$(MajorVersion).$(MinorVersion).$(PatchVersion).$(BuildVersion)</Version>
+    <AssemblyVersion>$(MajorVersion).$(MinorVersion).$(PatchVersion)</AssemblyVersion>
```
Comment (visschersm) [inline src/Utilities.Mongo/Utilities.Mongo.csproj]: Does this solve the assembly version when included or do we need to change the action as well?
See #11 for more info.

File: tests/Utilities.Mongo.UnitTests/MongoContextTests.cs [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using MongoDB.Driver;
+using System;
+
+namespace MatrTech.Utilities.Mongo.UnitTests
+{
+    [TestClass]
+    public class MongoContextTests
+    {
+        private const string connectionUrl = "mongodb://localhost:27017";
```
Comment (visschersm) [inline tests/Utilities.Mongo.UnitTests/MongoContextTests.cs]: Needs some cleanup of the created databases

File: src/Utilities.Mongo/Interfaces/IMongoContext.cs (hunk 5 above)
Comment (visschersm) [reply src/Utilities.Mongo/Interfaces/IMongoContext.cs:6]: Will verify

Comment (visschersm) [reply src/Utilities.Mongo/Interfaces/IMongoContext.cs:6]: Will use interface summary:
image

Unless overridden by implementation:
image

File: tests/Utilities.Mongo.UnitTests/MongoContextTests.cs (hunk 10 above)
Comment (visschersm) [reply tests/Utilities.Mongo.UnitTests/MongoContextTests.cs]: Created issue #12

**Review by visschersm — COMMENTED** [review]
Resolved PR suggestions

**Review by pjotrcasteel — APPROVED** [review]

**Pull request merged**